hail/src/main/scala/is/hail/expr/ir/Emit.scala

```diff
@@ -1597,6 +1597,137 @@ private class Emit(
+
+        val elementMul = lPType.elementType match {
+          case PInt32(_) => (a: Code[_], b: Code[_]) => (coerce[Int](a) * coerce[Int](b)).asInstanceOf[Code[Any]]
+          case PInt64(_) => (a: Code[_], b: Code[_]) => (coerce[Long](a) * coerce[Long](b)).asInstanceOf[Code[Any]]
+          case PFloat32(_) => (a: Code[_], b: Code[_]) => (coerce[Float](a) * coerce[Float](b)).asInstanceOf[Code[Any]]
+          case PFloat64(_) => (a: Code[_], b: Code[_]) => (coerce[Double](a) * coerce[Double](b)).asInstanceOf[Code[Any]]
+        }
+
+        val elementAdd = lPType.elementType match {
+          case PInt32(_) => (a: Code[_], b: Code[_]) => (coerce[Int](a) + coerce[Int](b)).asInstanceOf[Code[Any]]
+          case PInt64(_) => (a: Code[_], b: Code[_]) => (coerce[Long](a) + coerce[Long](b)).asInstanceOf[Code[Any]]
+          case PFloat32(_) => (a: Code[_], b: Code[_]) => (coerce[Float](a) + coerce[Float](b)).asInstanceOf[Code[Any]]
+          case PFloat64(_) => (a: Code[_], b: Code[_]) => (coerce[Double](a) + coerce[Double](b)).asInstanceOf[Code[Any]]
+        }
+
+        val elementZero = lPType.elementType match {
+          case PInt32(_) => const(0).asInstanceOf[Code[Any]]
+          case PInt64(_) => const(0L).asInstanceOf[Code[Any]]
+          case PFloat32(_) => const(0.0f).asInstanceOf[Code[Any]]
+          case PFloat64(_) => const(0.0).asInstanceOf[Code[Any]]
+        }
+
+        val eVti = typeToTypeInfo(lPType.elementType.virtualType)
+
+        val emitter = new NDArrayEmitter(mb, outputPType.nDims, unifiedShapeArray, lPType.shape.pType, lPType.elementType, setup) {
+          override def outputElement(idxVars: Array[Code[Long]]): Code[_] = {
+            val seqIdxVars = idxVars.toSeq
+            val element = coerce[Any](mb.newField("matmul_element")(eVti))
+            val kLocal = mb.newLocal[Long]
+            val k = kLocal.load()
+
+            val (lIdxVars: Array[Code[Long]], rIdxVars: Array[Code[Long]]) = (lNDims, rNDims) match {
+              case (1, 1) => (Array(k), Array(k))
+              case (1, _) =>
+                val stackDims :+ m = seqIdxVars
+
+                val rStackVars = NDArrayEmitter.zeroBroadcastedDims(stackDims.toArray, rightBroadcastFlags)
+                (Array(k), rStackVars :+ k :+ m)
+              case (_, 1) =>
+                val stackDims :+ n = seqIdxVars
+
+                val lStackVars = NDArrayEmitter.zeroBroadcastedDims(stackDims.toArray, leftBroadcastFlags)
+                (lStackVars :+ n :+ k, Array(k))
+              case (_, _) => {
+                val stackDims :+ n :+ m = seqIdxVars
+                val lStackVars = NDArrayEmitter.zeroBroadcastedDims(stackDims.toArray, leftBroadcastFlags)
+                val rStackVars = NDArrayEmitter.zeroBroadcastedDims(stackDims.toArray, rightBroadcastFlags)
+
+                (lStackVars :+ n :+ k, rStackVars :+ k :+  m)
+              }
+            }
```

Comment:
I think this is simpler as:
```scala
  val (lIndices, rIndices) = (leftShapeArray.length, rightShapeArray.length, idxVars.toSeq) match {
    case (1, 1, Seq()) =>
      (Seq(k), Seq(k))
    case (1, _, stack :+ m) =>
      (Seq(k), stack :+ k :+ m)
    case (_, 1, stack) =>
      (stack :+ k, Seq(k))
    case (_, _, stack :+ m) =>
      (stack :+ m :+ k, stack :+ k :+ m)
    case (a, b, c) =>
      fatal(s"bad mat mul: $a $b $c")
  }
  def broadcast(indices: Array[Code[Long]], shapes: Array[Code[Long]]): Array[Code[Long]] =
    (indices zip shapes).map { case (index, shape) => (shape ceq 0L).mux(0L, index) }

  val lIndicesBroadcasted = broadcast(lIndices, leftShapeArray)
  val rIndicesBroadcasted = broadcast(rIndices, rightShapeArray)
```

I know this recomputes the `shape ceq 0L` possibly multiple times but I hope the JIT can handle it and it's much easier to follow the intuition here.

Reply:
I simplified a bit to match the the three way match thing you're doing here, but I didn't switch to the inline broadcast. This doesn't quite work because I'm not zipping all the indices with the shape. Only the things that are in `stack` in the code snippet actually get broadcasted. I also feel like computing the flags in advance is not that different for readibility. flags is a bad name for it though, I'm going to call it `mask` instead since it's basically a bitmask.